### PR TITLE
add unit tests for probe errors and ExecProbeTimeout

### PR DESCRIPTION
**Pull request by nilskch**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This PR adds some unit tests.

#### Special notes for your reviewer:
I love k8s and I cloned the repo today for the first time and had a first look on the code. I would love to really contribute in the future, but this is my first contribution to a big open source project ever. Please accept my PR :)

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/probe/exec/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (wzshiming) [inline pkg/probe/exec/errors_test.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: pkg/probe/exec/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+*/
+
+package exec
+
+import (
+	"errors"
+	"testing"
+	"time"
+)
+
```
Comment (wzshiming) [inline pkg/probe/exec/errors_test.go:21]: We can combine the 3 tests into one, then we can extend them more easily, like the following.

https://github.com/kubernetes/kubernetes/blob/4211826c3c835ab6940e5af536e01434feac8d30/pkg/probe/exec/exec_test.go#L103-L148

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/probe/exec/errors_test.go (hunk 2 above)
Comment (nilskch) [reply pkg/probe/exec/errors_test.go:21]: Okay I will do this.

**Comment by nilskch** [comment]
@wzshiming can I change one function in the following way?

Instead of:

```go
func (t *TimeoutError) Error() string {
	return t.err.Error()
}
```
I would suggest this:
```go
func (t *TimeoutError) Error() string {
	if t.err == nil {
		return ""
	}
	return t.err.Error()
}
```

This prevents a nil pointer dereference when the error is nil. I would write the tests to cover this as well.

**Comment by bart0sh** [comment]
@nilskch please, fix failing CI tests, thanks.

**Comment by nilskch** [comment]
@bart0sh I fixed the failing CI test and combined the three unit tests into one as @wzshiming suggested.

**Comment by bart0sh** [comment]
@nilskch There is one [uncovered line of code](https://github.com/kubernetes/kubernetes/blob/master/pkg/probe/exec/exec.go#L79) related to the TimeoutError. It would be great if you can modify existing tests to cover it.

File: pkg/probe/exec/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			err:     errors.New("some error message"),
+			timeout: time.Hour * 8,
+			message: "some error message",
+		},
+		{
+			err:     errors.New("different error"),
+			timeout: time.Second * 61,
+			message: "different error",
+		},
```
Comment (bart0sh) [inline pkg/probe/exec/errors_test.go:41]: How this test case is different from the previous one?

File: pkg/probe/exec/errors_test.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			err:     errors.New(strings.Repeat("long-err", 128*10)),
+			timeout: time.Hour * 5,
+			message: strings.Repeat("long-err", 128*10),
+		},
```
Comment (bart0sh) [inline pkg/probe/exec/errors_test.go:46]: What does this test case test that is not tested by the previous test cases?

File: pkg/probe/exec/errors_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	for i, test := range tests {
+		testErr := NewTimeoutError(test.err, test.timeout)
+
+		if testErr == nil {
+			t.Errorf("[%d] expected error a TimeoutError, got nil", i)
+		}
+		if msg := testErr.Error(); msg != test.message {
+			t.Errorf("[%d] expected error message \"%s\", got \"%s\"", i, test.message, msg)
```
Comment (wzshiming) [inline pkg/probe/exec/errors_test.go:56]:
```suggestion
			t.Errorf("[%d] expected error message %q, got %q", i, test.message, msg)
```

File: pkg/probe/exec/errors_test.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		}
+		if timeout := testErr.Timeout(); timeout != test.timeout {
+			t.Errorf("[%d] expected timeout \"%s\", got \"%s\"", i, test.timeout, timeout)
```
Comment (wzshiming) [inline pkg/probe/exec/errors_test.go:59]:
```suggestion
			t.Errorf("[%d] expected timeout %q, got %q", i, test.timeout, timeout)
```

**Comment by nilskch** [comment]
@bart0sh the test coverage is now at 100%. I did not test the log statements, I hope this is ok. I also removed unnecessary tests.

File: pkg/probe/exec/errors_test.go (hunk 3 above)
Comment (nilskch) [reply pkg/probe/exec/errors_test.go]: you are right, I removed this test case.

File: pkg/probe/exec/errors_test.go (hunk 4 above)
Comment (nilskch) [reply pkg/probe/exec/errors_test.go]: I removed this test case as well.

**Comment by pacoxu** [comment]
/retitle add unit tests for probe errors and ExecProbeTimeout

**Comment by bart0sh** [comment]
@nilskch please, squash your commits into one, thanks

**Comment by nilskch** [comment]
One test failed with Pod scheduling timeout.
/test pull-kubernetes-verify

**Comment by derekwaynecarr** [comment]
Thank you for the test.

looks good.

/lgtm
/approve

**Pull request merged**